### PR TITLE
Use content-length header in ASGI instead of reading request body

### LINKED ISSUE #1573: Problem accessing to the request body after configure Sentry in FastAPI
### Steps to Reproduce

1. Create a Middleware to log the request body like https://stackoverflow.com/questions/70134618/why-does-fastapi-hang-when-adding-a-middleware-to-print-the-http-request-body 
2. The previous step works before configuring the Sentry SDK
3. Configure the Sentry as defined in https://docs.sentry.io/platforms/python/guides/fastapi/
4. When try to access to the test URL, the application hangs because the request body was already accessed by the Sentry integration

### Expected Result

Simplified middleware example

    @app.middleware("http")
    async def logging_middleware(request: Request, call_next):
        start_time = time.time()
        body = await request.body()
        print(body)
        response = await call_next(request)`
        return response

### Actual Result

When the Sentry implementation is active, the `body = await request.body()` line hangs and never response.

### Additional Info

This answer has more information related to the problem https://stackoverflow.com/a/71811390/11030205

### LINKED ISSUE #1595: FastAPI file upload not working with sentry after 1.9.6
### How do you use Sentry?

Sentry Saas (sentry.io)

### Version

1.9.6

### Steps to Reproduce

```
from fastapi import FastAPI, File, UploadFile
from PIL import Image
import uvicorn
import io

import sentry_sdk
from sentry_sdk.integrations.starlette import StarletteIntegration
from sentry_sdk.integrations.fastapi import FastApiIntegration

sentry_sdk.init(
    dsn=SECRET,
    release="server",
    integrations=[
        StarletteIntegration(),
        FastApiIntegration(),
    ],
    attach_stacktrace=True,
    request_bodies="always",
    send_default_pii=True,
    traces_sample_rate=0.0,
)

app = FastAPI()


@app.post("/predict")
async def predict(file: UploadFile = File(default=None)):
    contents = await file.read()
    Image.open(io.BytesIO(contents)).convert("RGB")
    return {"result": "ok"}


uvicorn.run(app, host="0.0.0.0", port=8888, loop="asyncio", lifespan="on")
```

### Expected Result

```{"result": "ok"}```

### Actual Result

```RuntimeError: The Sentry Python SDK can now automatically support ASGI frameworks like Starlette and FastAPI.
Please remove 'SentryAsgiMiddleware' from your project.
See https://docs.sentry.io/platforms/python/guides/asgi/ for more information.
```

so, when disabling sentry - all works fine, when adding sentry - it breaks. apparently sentry messes up content of uploaded file.

### LINKED ISSUE #1631: Upgrading from 1.9.5 to 1.9.8 triggers OOM error
### How do you use Sentry?

Sentry Saas (sentry.io)

### Version

1.9.8

### Steps to Reproduce

hello, fastapi app here with an upload endpoint, files uploaded are usually 15Mo in size, 

reverting to 1.9.5 fixes the issue, on 1.9.8 it's exhausting the container memory quite fast on an upload of several files (about 8 generally).

I'm using sentry this way, just reading the notes I see a deprecation, maybe the reason of the crash

```
    if app_settings_loaded.sentry_dsn is not None:
        # sentry middleware
        sentry_sdk.init(dsn=app_settings_loaded.sentry_dsn)
        app.add_middleware(SentryAsgiMiddleware)
```
that said I dont see the deprecation warning in logs.

the app crash and I dont get sentry logs, I think it's normal considering it appears to be an issue with it

### Expected Result

well to be able to upload my files :)

### Actual Result

```
backend_1          | INFO:     172.18.0.3:54200 - 2022-09-19 13:57:21,244 -'POST /uploads/x1 HTTP/1.1' 500 Internal Server Error
backend_1          | ERROR:    Exception in ASGI application
backend_1          | Traceback (most recent call last):
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 404, in run_asgi
backend_1          |     result = await app(  # type: ignore[func-returns-value]
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
backend_1          |     return await self.app(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/fastapi/applications.py", line 269, in __call__
backend_1          |     await super().__call__(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 293, in _sentry_patched_asgi_app
backend_1          |     return await middleware(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 138, in _run_asgi3
backend_1          |     return await self._run_app(scope, lambda: self.app(scope, receive, send))
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 187, in _run_app
backend_1          |     raise exc from None
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 182, in _run_app
backend_1          |     return await callback()
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/applications.py", line 124, in __call__
backend_1          |     await self.middleware_stack(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 98, in _create_span_call
backend_1          |     await old_call(*args, **kwargs)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 184, in __call__
backend_1          |     raise exc
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/middleware/errors.py", line 162, in __call__
backend_1          |     await self.app(scope, receive, _send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 98, in _create_span_call
backend_1          |     await old_call(*args, **kwargs)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/opentelemetry/instrumentation/asgi/__init__.py", line 346, in __call__
backend_1          |     await self.app(scope, wrapped_receive, wrapped_send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 138, in _run_asgi3
backend_1          |     return await self._run_app(scope, lambda: self.app(scope, receive, send))
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 148, in _run_app
backend_1          |     raise exc from None
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/asgi.py", line 145, in _run_app
backend_1          |     return await callback()
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/middleware/cors.py", line 92, in __call__
backend_1          |     await self.simple_response(scope, receive, send, request_headers=headers)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/middleware/cors.py", line 147, in simple_response
backend_1          |     await self.app(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/ratelimit/core.py", line 82, in __call__
backend_1          |     return await self.app(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 191, in _sentry_exceptionmiddleware_call
backend_1          |     await old_call(self, scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 191, in _sentry_exceptionmiddleware_call
backend_1          |     await old_call(self, scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 98, in _create_span_call
backend_1          |     await old_call(*args, **kwargs)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/exceptions.py", line 93, in __call__
backend_1          |     raise exc
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/exceptions.py", line 82, in __call__
backend_1          |     await self.app(scope, receive, sender)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 98, in _create_span_call
backend_1          |     await old_call(*args, **kwargs)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
backend_1          |     raise e
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
backend_1          |     await self.app(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/routing.py", line 670, in __call__
backend_1          |     await route.handle(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/routing.py", line 266, in handle
backend_1          |     await self.app(scope, receive, send)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/routing.py", line 65, in app
backend_1          |     response = await func(request)
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/fastapi.py", line 77, in _sentry_app
backend_1          |     info = await extractor.extract_request_info()
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 443, in extract_request_info
backend_1          |     content_length = await self.content_length()
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 480, in content_length
backend_1          |     raw_data = await self.raw_data()
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/sentry_sdk/integrations/starlette.py", line 491, in raw_data
backend_1          |     return await self.request.body()
backend_1          |   File "/opt/venv/lib/python3.10/site-packages/starlette/requests.py", line 236, in body
backend_1          |     self._body = b"".join(chunks)
backend_1          | MemoryError
```

### LINKED ISSUE #1646: Enhance handing of ASGI request body
### Problem Statement

In ASGI the request body is a stream and once it has been consumed it can not be consumed again.

This leads to some problems because in case of an error we try to extract information from the request body to give more context (this is also happening for transaction data, I guess)

Related issues:
https://github.com/getsentry/sentry-python/issues/1631
https://github.com/getsentry/sentry-python/issues/1595
https://github.com/getsentry/sentry-python/pull/1632#issuecomment-1261020099

### Solution Brainstorm

Refactor the `RequestExtractor` of all frameworks that can run in ASGI servers. Also improve the `DjangoRequestExtractor` so it can handle both WSGI and ASGI requests.

Also make the `AnnotatedValue` more easily understandable, because now this is a lot of magic numbers that no one really know what they are doing. 

The length and range is not used in those values, so the minimum looks like this: `{ "rem": [["!config", "x"]] }`
And `!raw` means that the body was not parsable (so .json() and .parsed_body() did return None) and `!config` means the body was over a certain threshold and therefore too big for sending.

### PR DESCRIPTION
**Pull request by antonpirker**
We did read the whole request body to figuring out the size of the content. This is not optimal at all. Now we read the "content-length" request header and if this header is not there, we just do not touch the request body (because in ASGI the request body is a stream and if it is consumed it can not be accessed again.)

Fixes #1646 
Fixes #1631
Fixes #1595
Fixes #1573

### CODE REVIEW

**Comment by sl0thentr0py** [comment]
can you make the PR title more specific? something like 
> Use content-length header in ASGI instead of reading request body
This is what goes into the changelog and right now our changelogs are not very informative.

File: tests/integrations/bottle/test_bottle.py [hunk 1]
```diff
@@ -234,9 +234,7 @@ def index():
     assert response[1] == "200 OK"
 
     (event,) = events
-    assert event["_meta"]["request"]["data"] == {
-        "": {"len": 2000, "rem": [["!config", "x", 0, 2000]]}
-    }
+    assert event["_meta"]["request"]["data"] == {"": {"rem": [["!config", "x"]]}}
```
Comment (sl0thentr0py) [inline tests/integrations/bottle/test_bottle.py:237]: did you check that ingestion does not fail with sending less fields than before?

File: sentry_sdk/integrations/starlette.py [hunk 2]
```diff
@@ -447,26 +447,16 @@ async def extract_request_info(self):
...
 
-            if not request_body_within_bounds(client, content_length):
-                data = AnnotatedValue(
-                    "",
-                    {
-                        "rem": [["!config", "x", 0, content_length]],
-                        "len": content_length,
-                    },
-                )
+            if not content_length or not request_body_within_bounds(
```
Comment (sl0thentr0py) [inline sentry_sdk/integrations/starlette.py:450]: if `content_length` is not there doesn't mean that we truncated it.
This should be a separate case. We should just not do anything in that case.
So
```python
if content_length:
   if not request_body_bla():
      data = AnnotatedValue
    else:
       do stuff
```

**Review by sl0thentr0py — CHANGES_REQUESTED** [review]

File: tests/integrations/bottle/test_bottle.py (hunk 1 above)
Comment (antonpirker) [reply tests/integrations/bottle/test_bottle.py:237]: Yes. Joris said here that those fields are optional: https://github.com/getsentry/sentry-python/issues/1646 (see the slack screenshots)

Comment (antonpirker) [reply tests/integrations/bottle/test_bottle.py:237]: Ok, I did not check myself. I trust Joris...

File: sentry_sdk/integrations/starlette.py (hunk 2 above)
Comment (antonpirker) [reply sentry_sdk/integrations/starlette.py]: done.

**Review by sl0thentr0py — APPROVED** [review]
🚢

**Pull request merged**